File: engine/static/input/src/mouse.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/macros.hpp"
#include "argus/lowlevel/math.hpp"

#include "argus/wm/window.hpp"

#include "argus/input/input_manager.hpp"
#include "argus/input/mouse.hpp"
#include "internal/input/event_helpers.hpp"
#include "internal/input/mouse.hpp"
#include "internal/input/pimpl/controller.hpp"
#include "internal/input/pimpl/input_manager.hpp"

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wold-style-cast"
#include "SDL_events.h"
#pragma GCC diagnostic pop
#include "SDL_mouse.h"

#include <mutex>
#include <string>
#include <unordered_map>

namespace argus::input {
    static const std::unordered_map<MouseButton, int> g_mouse_button_mappings({
            { MouseButton::Primary, SDL_BUTTON_LEFT },
            { MouseButton::Middle, SDL_BUTTON_MIDDLE },
            { MouseButton::Secondary, SDL_BUTTON_RIGHT },
            { MouseButton::Back, SDL_BUTTON_X1 },
            { MouseButton::Forward, SDL_BUTTON_X2 },
    });

    argus::Vector2d mouse_pos(void) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        return InputManager::instance().pimpl->mouse_state.last_pos;
    }

    argus::Vector2d mouse_delta(void) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        return InputManager::instance().pimpl->mouse_state.delta;
    }

    double get_mouse_axis(MouseAxis axis) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        switch (axis) {
            case MouseAxis::Horizontal:
                return InputManager::instance().pimpl->mouse_state.last_pos.x;
            case MouseAxis::Vertical:
                return InputManager::instance().pimpl->mouse_state.last_pos.y;
            default:
                throw std::invalid_argument("Unknown mouse axis ordinal " + std::to_string(int(axis)));
        }
    }

    double get_mouse_axis_delta(MouseAxis axis) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        auto &state = InputManager::instance().pimpl->mouse_state;

        double val;
        switch (axis) {
            case MouseAxis::Horizontal:
                val = state.delta.x;
                break;
            case MouseAxis::Vertical:
                val = state.delta.y;
                break;
            default:
                throw std::invalid_argument("Unknown mouse axis ordinal " + std::to_string(int(axis)));
        }
        return val;
    }

    bool is_mouse_button_pressed(MouseButton button) {
        if (int(button) < 0) {
            throw std::invalid_argument("Invalid mouse button ordinal " + std::to_string(int(button)));
        }
        auto sdl_button = g_mouse_button_mappings.find(button);
        if (sdl_button == g_mouse_button_mappings.cend()) {
            throw std::invalid_argument("Invalid mouse button ordinal " + std::to_string(int(button)));
        }

        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        return (InputManager::instance().pimpl->mouse_state.button_state & SDL_BUTTON(sdl_button->second)) != 0;
    }

    static void _poll_mouse(void) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);

        auto &state = InputManager::instance().pimpl->mouse_state;

        int x;
        int y;
        state.button_state = SDL_GetMouseState(&x, &y);

        if (state.got_first_pos) {
            state.delta.x += double(x) - state.last_pos.x;
            state.delta.y += double(y) - state.last_pos.y;
        } else {
            state.got_first_pos = true;
        }

        state.last_pos.x = double(x);
        state.last_pos.y = double(y);
    }

    static void _dispatch_button_event(const Window &window, MouseButton button, bool release) {
        for (auto &pair : InputManager::instance().pimpl->controllers) {
            auto controller_index = pair.first;
            auto &controller = *pair.second;

            auto it = controller.pimpl->mouse_button_to_action_bindings.find(button);
            if (it == controller.pimpl->mouse_button_to_action_bindings.end()) {
                continue;
            }

            for (auto &action : it->second) {
                dispatch_button_event(window, controller_index, action, release);
            }
        }
    }

    static void _dispatch_axis_events(const Window &window, double x, double y, double dx, double dy) {
        for (auto &pair : InputManager::instance().pimpl->controllers) {
            auto controller_index = pair.first;
            auto &controller = *pair.second;

            auto it_x = controller.pimpl->mouse_axis_to_action_bindings.find(MouseAxis::Horizontal);
            auto it_y = controller.pimpl->mouse_axis_to_action_bindings.find(MouseAxis::Vertical);
            if (it_x != controller.pimpl->mouse_axis_to_action_bindings.end()) {
                for (auto &action : it_x->second) {
                    dispatch_axis_event(window, controller_index, action, x, dx);
                }
            }

            if (it_y == controller.pimpl->mouse_axis_to_action_bindings.end()) {
                for (auto &action: it_y->second) {
                    dispatch_axis_event(window, controller_index, action, y, dy);
                }
            }
        }
    }

    static void _handle_mouse_events(void) {
        constexpr size_t event_buf_size = 8;
        SDL_Event events[event_buf_size];

        int to_process;
        while ((to_process = SDL_PeepEvents(events, event_buf_size,
                SDL_GETEVENT, SDL_MOUSEMOTION, SDL_MOUSEBUTTONUP)) > 0) {
            for (int i = 0; i < to_process; i++) {
                auto &event = events[i];

                auto *window = get_window_from_handle(SDL_GetWindowFromID(event.key.windowID));
                if (window == nullptr) {
                    return;
                }

                if (event.type == SDL_MOUSEMOTION) {
                    _dispatch_axis_events(*window, event.motion.x, event.motion.y,
                            event.motion.xrel, event.motion.yrel);
                } else {
                    MouseButton button;
                    switch (event.button.button) {
                        case SDL_BUTTON_LEFT:
                            button = MouseButton::Primary;
                            break;
                        case SDL_BUTTON_RIGHT:
                            button = MouseButton::Secondary;
                            break;
                        case SDL_BUTTON_MIDDLE:
                            button = MouseButton::Middle;
                            break;
                        case SDL_BUTTON_X1:
                            button = MouseButton::Back;
                            break;
                        case SDL_BUTTON_X2:
                            button = MouseButton::Forward;
                            break;
                        default:
                            Logger::default_logger().debug("Ignoring unrecognized mouse button with ordinal %d",
                                    event.button.which);
                            return;
                    }
                    _dispatch_button_event(*window, button,
                            event.type == SDL_MOUSEBUTTONUP);
                }
            }
        }
    }

    void init_mouse(const argus::Window &window) {
        UNUSED(window);
        // no-op
    }

    void update_mouse(void) {
        _poll_mouse();
        _handle_mouse_events();
    }

    void flush_mouse_delta(void) {
        std::lock_guard<std::mutex> lock(InputManager::instance().pimpl->mouse_state_mutex);
        InputManager::instance().pimpl->mouse_state.delta = {};
    }
}

File: engine/static/input_rs/src/mouse.rs
use std::collections::HashMap;
use lazy_static::lazy_static;
use wm_rustabi::argus::wm::{get_window_from_handle, Window};
use argus_scripting_bind::script_bind;
use lowlevel_rustabi::argus::lowlevel::Vector2d;
use sdl2::events::{sdl_get_events, SdlEventData, SdlEventType};
use sdl2::mouse::{sdl_get_mouse_state, SdlMouseButton};
use sdl2::video::SdlWindow;
use crate::input_event::{dispatch_axis_event, dispatch_button_event};
use crate::InputManager;

lazy_static! {
    static ref MOUSE_BUTTON_MAP_ARGUS_TO_SDL: HashMap<MouseButton, SdlMouseButton> = HashMap::from([
        (MouseButton::Primary, SdlMouseButton::Left),
        (MouseButton::Middle, SdlMouseButton::Middle),
        (MouseButton::Secondary, SdlMouseButton::Right),
        (MouseButton::Back, SdlMouseButton::X1),
        (MouseButton::Forward, SdlMouseButton::X2),
    ]);
    static ref MOUSE_BUTTON_MAP_SDL_TO_ARGUS: HashMap<SdlMouseButton, MouseButton> =
        MOUSE_BUTTON_MAP_ARGUS_TO_SDL.iter().map(|(k, v)| (*v, *k)).collect();
}

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[script_bind]
pub enum MouseButton {
    Primary = 1,
    Secondary = 2,
    Middle = 3,
    Back = 4,
    Forward = 5,
}

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[script_bind]
pub enum MouseAxis {
    Horizontal,
    Vertical,
}

/// Returns the change in position of the mouse from the previous frame.
#[script_bind]
pub fn mouse_delta() -> Vector2d {
    let mouse_state = InputManager::instance().mouse_state.read();
    mouse_state.delta
}

/// Gets the current position of the mouse within the currently focused window.
#[script_bind]
pub fn mouse_pos() -> Vector2d {
    let mouse_state = InputManager::instance().mouse_state.read();
    mouse_state.last_pos.unwrap_or_default()
}

// Gets the current value of a mouse axis.
#[script_bind]
pub fn get_mouse_axis(axis: &MouseAxis) -> f64 {
    let mouse_state = InputManager::instance().mouse_state.read();
    match axis {
        MouseAxis::Horizontal => mouse_state.last_pos.map(|v| v.x),
        MouseAxis::Vertical => mouse_state.last_pos.map(|v| v.y),
    }.unwrap_or_default()
}

/// Gets the last delta of a mouse axis.
#[script_bind]
pub fn get_mouse_axis_delta(axis: &MouseAxis) -> f64 {
    let mouse_state = InputManager::instance().mouse_state.read();
    match axis {
        MouseAxis::Horizontal => mouse_state.delta.x,
        MouseAxis::Vertical => mouse_state.delta.y,
    }
}

/// Gets whether a mouse button is currently being pressed.
#[script_bind]
pub fn is_mouse_button_pressed(button: &MouseButton) -> bool {
    let Some(sdl_button) = MOUSE_BUTTON_MAP_ARGUS_TO_SDL.get(button) else {
        panic!("Invalid mouse button {:?}", button);
    };
    let mouse_state = InputManager::instance().mouse_state.read();
    (mouse_state.button_state & sdl_button.get_mask() as u32) != 0
}

fn poll_mouse() {
    let mut mouse_state = InputManager::instance().mouse_state.write();

    let new_state = sdl_get_mouse_state();
    let new_x = new_state.position_x;
    let new_y = new_state.position_y;

    if let Some(last_pos) = mouse_state.last_pos {
        mouse_state.delta.x += new_x as f64 - last_pos.x;
        mouse_state.delta.y += new_y as f64 - last_pos.y;
    }

    mouse_state.last_pos = Some(Vector2d::new(new_x as f64, new_y as f64))
}

fn dispatch_mouse_button_event(
    window: &Window,
    button: &MouseButton,
    release: bool
) {
    for item in &InputManager::instance().controllers {
        let controller_name = item.key();
        let controller = item.value();

        let Some(actions) = controller.get_mouse_button_actions(button) else {
            continue;
        };

        for action in actions {
            dispatch_button_event(Some(window), controller_name.clone(), action, release);
        }
    }
}

fn dispatch_mouse_axis_events(window: &Window, x: f64, y: f64, dx: f64, dy: f64) {
    for item in &InputManager::instance().controllers {
        let controller_name = item.key();
        let controller = item.value();
        if let Some(actions_h) = controller.get_mouse_axis_actions(&MouseAxis::Horizontal) {
            for action in actions_h {
                dispatch_axis_event(Some(&window), controller_name.clone(), action, x, dx);
            }
        }

        if let Some(actions_v) = controller.get_mouse_axis_actions(&MouseAxis::Vertical) {
            for action in actions_v {
                dispatch_axis_event(Some(&window), controller_name.clone(), action, y, dy);
            }
        }
    }
}

fn handle_mouse_events() {
    for event in sdl_get_events(SdlEventType::MouseMotion, SdlEventType::MouseButtonUp) {
        match event.data {
            SdlEventData::MouseMotion(data) => {
                let Some(window) = (match SdlWindow::from_id(data.window_id) {
                    Ok(sdl_window) => get_window_from_handle(sdl_window.get_handle()),
                    Err(_) => None,
                }) else { continue; };
                dispatch_mouse_axis_events(
                    &window,
                    data.x as f64,
                    data.y as f64,
                    data.x_rel as f64,
                    data.y_rel as f64,
                );
            }
            SdlEventData::MouseButton(data) => {
                let Some(button) = MOUSE_BUTTON_MAP_SDL_TO_ARGUS.get(&data.button) else {
                    println!("Ignoring unrecognized mouse button {:?}", data.button);
                    return;
                };
                let Some(window) = (match SdlWindow::from_id(data.window_id) {
                    Ok(sdl_window) => get_window_from_handle(sdl_window.get_handle()),
                    Err(_) => None,
                }) else { continue; };
                dispatch_mouse_button_event(
                    &window,
                    button,
                    event.ty == SdlEventType::MouseButtonUp,
                );
            }
            _ => {}
        }
    }
}

pub(crate) fn init_mouse(_window: &Window) {
    // no-op
}

pub(crate) fn update_mouse() {
    poll_mouse();
    handle_mouse_events();
}

pub(crate) fn flush_mouse_delta() {
    let mut mouse_state = InputManager::instance().mouse_state.write();
    mouse_state.delta = Vector2d::new(0.0, 0.0);
}
